### ISSUE TITLE
Feature request: allow `order:dfs` + `uniqueVertices:global` and allow `uniqueEdges:global`

### ISSUE THREAD

**Issue by OlivierSohn** [arangodb/arangodb#15537]
Hello,

Currently, `order:dfs` + `uniqueVertices:global` is not allowed. [The doc](https://www.arangodb.com/docs/stable/aql/graphs-traversals.html) explains that it is not allowed because it would return unpredictable results.

However, `order:bfs` + `uniqueVertices:global` is allowed, and leads to non-deterministic results, like [the doc](https://www.arangodb.com/docs/stable/aql/graphs-traversals.html) explains:

> Note: Using this configuration the result is not deterministic any more. If there are multiple paths from startVertex to vertex, one of those is picked. In case of a weighted traversal, the path with the lowest weight is picked, but in case of equal weights it is undefined which one is chosen.

I would like to understand why one is allowed and not the other.

I understand that results of `order:dfs` + `uniqueVertices:global` will depend on the order of edges, but this the same for `order:bfs` + `uniqueVertices:global`.

I can imagine that a depth-first traversal will use less memory than a breadth-first traversal so allowing `order:dfs` + `uniqueVertices:global` may optimize some applications.

Hence, would it be possible to allow `order:dfs` + `uniqueVertices:global` in a future release? Or is there something I am missing that makes the `bfs` case fundamentally different from the `dfs` case? 

Also, would it be possible to allow `uniqueEdges:global`? There are applications where using `uniqueEdges:global` would be useful. Results will also be non deterministic but are useful in some cases.

Thank you!

**Comment by goedderz** [comment]
Hi @OlivierSohn!

I'll try to fill in some of the details.

> I would like to understand why one is allowed and not the other.

First, regarding the (non-)determinisim of `{order: "bfs", uniqueVertices: "global"}`. While the paths that are taken aren't deterministic, the set of vertices visited/returned will be the same. So regarding `FOR v, e, p IN ...`, the set of `v`s will be deterministic (and in any bfs-order). The set of `e`s and `p`s can be different though, that's what mentioned in the docs.

Now generally, that is similar for a `dfs`, which you correctly pointed out. But note that a traversal always has a maximum depth assigned, e.g. `FOR v, e, p IN min..max`. And with that and `uniqueVertices: "global"`, `dfs` behaves badly. Take for example a graph with the vertices `{s, a, b, c}`, and the edges `{(s, a), (s, b), (b, a), (a, c)}`.

![image](https://user-images.githubusercontent.com/11504143/149756537-7c335506-873b-460d-ab8b-42106ed43413.png)

A traversal `FOR v IN 0..2 OUTBOUND s`, with `dfs` and `uniqueVertices: "global"`, could either return `[s, a, c, b]` or `[s, b, a]`. This is because, when visiting `b` before `a`, `c` will be reached at depth 2.

So in short, for a `dfs` with `uniqueVertices: "global"` (and a maximum depth, which is not optional), not even the set of vertices returned is deterministic. That's why the combination is disallowed.

That being said, I get that this can be desirable. However, it will need more profound changes to AQL traversals than just allowing the option combination, so it can be a sensible choice.

> I can imagine that a depth-first traversal will use less memory than a breadth-first traversal

Not generally. The difference between a dfs and a bfs is the central data structure: a stack for dfs, and a queue for bfs. Depending on the actual data, one might use less memory than the other; but in the end, it's always linear, in particular for `uniqueVertices: "global"`, where the already visited vertices have to be kept track of.

> Also, would it be possible to allow uniqueEdges:global? There are applications where using uniqueEdges:global would be useful.

We are aware of that, and we have it in our backlog to discuss this. If you could describe your usecase(s) for this, it might be helpful.

Best regards,

Tobias

**Comment by OlivierSohn** [comment]
> So in short, for a `dfs` with `uniqueVertices: "global"` (and a maximum depth, which is not optional), not even the set of vertices returned is deterministic. That's why the combination is disallowed.

Thank you @goedderz ! Indeed it makes sense with your explanation.

> That being said, I get that this can be desirable. However, it will need more profound changes to AQL traversals than just allowing the option combination, so it can be a sensible choice.

In fact, knowing now that it could give different results I don't think it would use this functionality even if it was slightly faster than bfs.

> > I can imagine that a depth-first traversal will use less memory than a breadth-first traversal
> 
> Not generally. The difference between a dfs and a bfs is the central data structure: a stack for dfs, and a queue for bfs. Depending on the actual data, one might use less memory than the other; but in the end, it's always linear, in particular for `uniqueVertices: "global"`, where the already visited vertices have to be kept track of.

That is true, thanks.

> > Also, would it be possible to allow uniqueEdges:global? There are applications where using uniqueEdges:global would be useful.
> 
> We are aware of that, and we have it in our backlog to discuss this. If you could describe your usecase(s) for this, it might be helpful.

Cool! The use case I have in mind here is the following:
- I have an edge collection `Edges`
- I have a condition `Cond` (a filter) that tells me whether I am allowed to use a given edge of `Edges` or not, based on its properties
- I have a source vertex `S`
- a distance `Dist`
- I define the (abstract) subgraph `Sub` = all vertices `V` such that "distance between S and V using edges of `Edges` that meet `Cond`" <= `Dist`
- And I want a way to retrieve vertices / edges of `Sub`, with a reasonable complexity, ideally it should be linear in number of vertices + number of nodes of `Sub`.

The context here is that I need to compute all shortest paths between 2 nodes, and cannot use https://www.arangodb.com/docs/3.8/aql/graphs-k-paths.html because of the condition on edges of the edge collection.

**Comment by OlivierSohn** [comment]
Hello @goedderz ,

Thanks for your help!

> First, do a weighted traversal (with `{uniqueVertices: "global", order: "weighted"}`) to extract only the vertices of your desired subgraph. Combine it with PRUNE to stop when the maximum distance is exceeded (you will have to post-filter those).

In the PRUNE condition, can I also check for `Cond` to eliminate paths that would use forbidden edges, or will this filtering be too aggressive and remove paths that would be valid? 

Here is an example to explain the problem:
- The source vertex is `S`
- There are 2 edges from `S` to `A`, one of the edges is forbidden

If the traversal first discovers the forbidden edge, given that we have `uniqueVertices: "global"`, will the second edge be considered or discarded because it goes to a vertex that has already been "pruned"?

Thanks!

**Comment by goedderz** [comment]
> In the PRUNE condition, can I also check for Cond to eliminate paths that would use forbidden edges, or will this filtering be too aggressive and remove paths that would be valid?

Don't do that in PRUNE; instead, add a `FILTER path.edges[*].color ALL == "red"` (assuming you want edges with `edge.color == "RED"`).

In PRUNE, check for something like `PRUNE path.edges[*].length > @Dist", assuming `length` provides the cost of an edge. `length` must be the same attribute as supplied via `weightAttribute` to the weighted traversal.

PRUNE works as follows: If the condition is `true` for some vertex `v`, it will still be returned by the traversal, but the traversal will stop there, i.e. no incident edges from `v` will be traversed. In combination with `uniqueVertices: "global"` that still means that the vertex was visited, and it will not be visited again from any other edge during this traversal. So as you suspected, if you'd check the edge condition during prune, this would lead to erroneous results.

**Comment by OlivierSohn** [comment]
@goedderz 

Yes that is a bit unfortunate. I was hoping to find a solution which "scales well" in case the graph is densely connected and the condition on edges is highly effective at reducing the number of traversed edges/nodes. I guess that the feature you have logged for the future (https://github.com/arangodb/arangodb/issues/15537#issuecomment-1055194008) may address that concern.

Thanks for the help!

**Comment by OlivierSohn** [comment]
@goedderz I didn't know that the optimizer would do this, so it sounds like this should be exactly what we need. I will forward the information to the team responsible of writing these AQL queries. Thanks again.

**Comment by goedderz** [comment]
Happy to help!
